File: LeetCode/Java/0876_Middle_of_the_Linked_List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode middleNode(ListNode head) {
        if (head.next == null)
            return head;
        
        ListNode slow = head;
        ListNode fast = head;
        
        while (true){
            if (fast.next == null)
                return slow;
            if (fast.next.next == null)
                return slow.next;
            
            slow = slow.next;
            fast = fast.next.next;
        }
    }
}

File: LeetCode/Python/0876_Middle_of_the_Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def middleNode(self, head: ListNode) -> ListNode:
        if not head.next:
            return head
        
        slow = fast = head
        while True:
            if not fast.next:
                return slow
            if not fast.next.next:
                return slow.next
            slow = slow.next
            fast = fast.next.next
